### DISCUSSION TITLE
C# - Identical entities - Derived instance - Sync script

### DISCUSSION THREAD

**Discussion by Generalec** [stride3d/stride#2191] [Q&A] [answered]
I have two entity which are identical. I have each separated sync script for ObjectA and ObjectB entity in project.

ObjectA  ---- > ObjectA.cs
```
    // Sync script - ObjectA - Assigned on entity - ObjectA
    public class ObjectA : ObjectBase
    {

    }
	
```
ObjectB  ---- > ObjectB.cs
```
   
    // Sync script - ObjectB - Assigned on entity - ObjectB
    public class ObjectB : ObjectBase
    {

    }
```

`	DebugText.Print($"{ObjectA.GetEntity.Name} - {ObjectA.GetEntity.Transform.Position}", new Int2(x: 20, y: 185));
    DebugText.Print($"{ObjectB.GetEntity.Name} - {ObjectB.GetEntity.Transform.Position}", new Int2(x: 20, y: 200));`

Output is:
```
ObjectB - 0;0;0
ObjectB - 0;0;0
```

Instead of:
```
ObjectA - 0;2;0
ObjectB - 0;0;0
```

Problem is that derived classes are"sharing" parrent instances (obviously).

Any suggestion how to solve this behaviour? Instead of coping content of ObjectBase to each script ?

Thank you in advance.

**Comment by manio143** [comment, marked as answer]
Can you share the code for your base class?
Component don't share state across instances unless you use static fields which then makes that state global.

**Comment by Generalec** [comment]
```
DebugText.Print($"{ObjectA.baseEntity.Name} - {ObjectA.baseEntity.Transform.Position}", new Int2(x: 20, y: 185)); 
DebugText.Print($"{ObjectB.baseEntity.Name} - {ObjectB.baseEntity.Transform.Position}", new Int2(x: 20, y: 200));
```

My code is horible and definitely needs some refactoring. I know I have private vars as static... Historical remains...

Yes. I use lot of static so if it's problem It's good to know what consequences are. 


I will rewrite to non static approach with events.

Clone of entity will not copy separate instance  of script, right ?

Thank you.

**Reply by manio143** [reply]
I believe a clone does have a new instance of the component, and only serializable parameters are preserved in the process.
Here I refer to the kind of clone which is the result of Prefab instantiation.

**Comment by Generalec** [comment]
@manio143 Thank you so much. Static variables of course were the issue.
